### ISSUE TITLE
[BUG] Exception running sync GetBlobs perf test

### ISSUE THREAD

**Issue by mikeharder** [Azure/azure-sdk-for-net#31570]
### Library name and version

12.13.0, main

### Describe the bug

When running a perf test with parameters `GetBlobs --count 500 --parallel 32 --sync`, we often (but not always) see the following exception.  This only repros when running the **sync** tests, it does not repro with **async**.  It repros on both 12.13.0 and main (only versions I've tested).

```
Unhandled exception. Unhandled exception. Unhandled exception. Unhandled exception.Unhandled exception.  Unhandled exception. Unhandled exception. Unhandled exception. System.NotSupportedException:  The ReadAsync method cannot be called when another read operation is pending.
   at System.Net.Security.SslStream.ReadAsyncInternal[TReadAdapter](TReadAdapter adapter, Memory`1 buffer)
   at System.Net.Security.SslStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.Net.Http.HttpConnection.Fill()
   at System.Net.Http.HttpConnection.ChunkedEncodingReadStream.Read(Span`1 buffer)
   at System.Net.Http.HttpBaseStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at Azure.Core.Pipeline.ResponseBodyPolicy.CopyTo(Stream source, Stream destination, CancellationTokenSource cancellationTokenSource)
...
   at Azure.Storage.Blobs.ContainerRestClient.ListBlobFlatSegment(String prefix, String marker, Nullable`1 maxresults, IEnumerable`1 include, Nullable`1 timeout, CancellationToken cancellationToken) in /home/mharder/net/sdk/storage/Azure.Storage.Blobs/src/Generated/ContainerRestClient.cs:line 1309
   at Azure.Storage.Blobs.BlobContainerClient.GetBlobsInternal(String marker, BlobTraits traits, BlobStates states, String prefix, Nullable`1 pageSizeHint, Boolean async, CancellationToken cancellationToken) in /home/mharder/net/sdk/storage/Azure.Storage.Blobs/src/BlobContainerClient.cs:line 2599
   at Azure.Storage.Blobs.Models.GetBlobsAsyncCollection.GetNextPageAsync(String continuationToken, Nullable`1 pageSizeHint, Boolean async, CancellationToken cancellationToken) in /home/mharder/net/sdk/storage/Azure.Storage.Blobs/src/Models/GetBlobsAsyncCollection.cs:line 71
   at Azure.Core.Pipeline.TaskExtensions.EnsureCompleted[T](ValueTask`1 task) in /home/mharder/net/sdk/core/Azure.Core/src/Shared/TaskExtensions.cs:line 53
   at Azure.Storage.StorageCollectionEnumerator`1.StoragePageable.GetEnumerator()+MoveNext() in /home/mharder/net/sdk/storage/Azure.Storage.Common/src/Shared/StorageCollectionEnumerator.cs:line 103
   at Azure.Storage.Blobs.Perf.Scenarios.GetBlobs.Run(CancellationToken cancellationToken) in /home/mharder/net/sdk/storage/Azure.Storage.Blobs/perf/Azure.Storage.Blobs.Perf/Scenarios/GetBlobs.cs:line 38
```

Example: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1887558&view=logs&j=42ded549-05ee-5cdb-7ba7-7a948a0cc056&t=49e32a0b-4245-531b-0723-9d45113ffbd3&l=8753

### Expected behavior

No exception

### Actual behavior

Exception

### Reproduction Steps

1. Create an Azure VM matching our perf agents
    * Size: D4ds_v5
    * OS: Ubuntu 20.04
    * Accelerated Networking: Enabled
2. In the same region, create a storage account
    * Kind: BlockBlobStorage
    * SKU: Premium_LRS
3. Run perf test on VM
    1. Clone repo
    2. Set env vars
    3. cd sdk/storage/Azure.Storage.Blobs/perf/Azure.Storage.Blobs.Perf
    4. dotnet run -c release -f netcoreapp3.1 -- GetBlobs --count 500 --parallel 32 --sync
    5. Re-run test until it fails with exception

**Comment by hpaul-osi** [comment]
I've encountered the same exception with GetBlobsAsync in 12.13.0. Not in the perf tests, but using this client to list blobs with different prefixes in parallel.

**Comment by mikeharder** [comment]
@hpaul-osi: Thanks for the report.

1. Can you share code and instructions to reproduce what you are seeing?
2. Does it repro consistently or intermittently?
3. Is this impacting a real-world scenario for you?

**Comment by hpaul-osi** [comment]
1. The code gets a count of most recent snapshots in a storage container, so it lists with GetBlobsAsync. The blobs all start with a hex prefix, so the listing of snapshots is done in parallel for each possible first character prefix with Parallel.ForEachAsync. 
2. These listings are done for several storage accounts, and we consistently see it for only one of those accounts which has 64 containers with approximately 50000 blobs in each, so your count may be to low to trigger it with your async performance tests.
3. 6.0.11
4. The scenario is real, but we caught this issue in dev environments, so it is not affecting production.

**Comment by mikeharder** [comment]
@hpaul-osi: I would like to try to reproduce your issue.  I can create 64 containers with 50k blobs each.  But can you share a code snippet showing how you are making the parallel calls to `GetBlobsAsync()`?  Is there a certain pattern to the container and blob names I should follow?

Coincidentally, we do run a performance test that creates a container with 50k blobs, and calls `GetBlobsAsync()` in 32 parallel threads.  We have never hit this exception in this test.  Only in our sync tests.  But I suspect this bug is a race condition so it might require specific circumstances to repro.

You said you caught this in dev but it's not impacting production.  Do you have a workaround for the issue in production?  Or is it that you just haven't hit this issue **yet** in production?

**Comment by hpaul-osi** [comment]
I would like to strip it down to the minimum to reliably reproduce the issue to share; I just haven't had time to produce the distilled case and verify it still reproduces the issue yet.

The issue is not impacting production because we detected this in lower environments before pushing the problematic code to production.

**Comment by mikeharder** [comment]
Thanks, no rush from our side.  We can already repro the exception reliably, so we might not need another repro.  But it's interesting that we can only repro with the sync api while you can repro with async.

**Comment by mikeharder** [comment]
We believe this can happen when any sync request is cancelled.

Simplified repro: https://github.com/mikeharder/get-blobs-cancel

We believe the bug is our code can call Dispose() concurrently with Read():

https://github.com/Azure/azure-sdk-for-net/blob/785c355bde40ae4b59d922611666f52b16074e6e/sdk/core/Azure.Core/src/Pipeline/Internal/ResponseBodyPolicy.cs#L80

Which can cause the "nested read" exception to be thrown:

https://github.com/dotnet/runtime/blob/6ddd06c023574f298f21f4acde58bf4af337d309/src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs#L55-L59

https://github.com/dotnet/runtime/blob/6ddd06c023574f298f21f4acde58bf4af337d309/src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs#L766-L769

**Comment by mikeharder** [comment]
Root cause: https://github.com/Azure/azure-sdk-for-net/issues/32577

Once the root cause is fixed, we should see if this still repros.

**Comment by mikeharder** [comment]
@hpaul-osi: So far, we have only been able to repro this exception by calling the sync API `GetBlobs(CancellationToken)`, and having the `CancellationToken` expire during the call.  We understand this issue (#32577) and will be making a fix.

We have not been able to repro this when calling the async API `GetBlobsAsync(CancellationToken)`, even if the `CancellationToken` expires during the call.  It may be **possible** to repro with the async API, but the race window might be much smaller.  We do believe the repro requires the request to be cancelled somehow, either programmatically or by hitting the `NetworkTimeout` (or perhaps a service/network error could trigger this).

So we would still be interested in your repro code, if you are able to repro using the async API.  Otherwise, once we fix #32577, you could try a pre-release build of `Azure.Core` to see if it fixes your issue.

**Comment by hpaul-osi** [comment]
I was unable to reproduce at all with a simple console app using the same listing logic. The dev environment with the service deployed still consistently fails. Reviewing the logs shows, these NotSupportedException exceptions are always accompanied by timeout cancellations, so it does appear likely it is the same issue, with a tighter window.
